File: frontend/src/api/knowledge.ts
import client from './client'

export interface PaginatedTable {
  columns: string[]
  rows: Record<string, any>[]
  total: number
  page: number
  page_size: number
}

export interface GeneDetail {
  gene: string
  sheets: Record<string, Record<string, any>[]>
}

export interface KBStats {
  gene_knowledge: { sheets: number; total_rows: number }
  drug_mappings: { total_rows: number }
  immune_genes: { total_rows: number }
}

export const knowledgeApi = {
  async getGenes(params: { search?: string; page?: number; page_size?: number } = {}): Promise<PaginatedTable> {
    const { data } = await client.get('/knowledge/genes', { params })
    return data.data
  },

  async getGeneDetail(geneName: string): Promise<GeneDetail> {
    const { data } = await client.get(`/knowledge/genes/${encodeURIComponent(geneName)}`)
    return data.data
  },

  async getDrugs(params: { search?: string; page?: number; page_size?: number } = {}): Promise<PaginatedTable> {
    const { data } = await client.get('/knowledge/drugs', { params })
    return data.data
  },

  async getImmuneGenes(): Promise<PaginatedTable> {
    const { data } = await client.get('/knowledge/immune-genes')
    return data.data
  },

  async getStats(): Promise<KBStats> {
    const { data } = await client.get('/knowledge/stats')
    return data.data
  },
}

File: backend/app/api/knowledge.py
"""Knowledge base browsing endpoints."""

from fastapi import APIRouter, Depends, Query

from app.dependencies import require_admin
from app.models.user import User
from app.schemas.common import ApiResponse
from app.services import knowledge_service as svc

router = APIRouter(prefix="/knowledge", tags=["knowledge"])


@router.get("/genes", response_model=ApiResponse)
def list_genes(
    search: str = Query("", description="Search keyword"),
    page: int = 1,
    page_size: int = 50,
):
    data = svc.get_gene_list(page=page, page_size=page_size, search=search)
    return ApiResponse(data=data)


@router.get("/genes/{gene_name}", response_model=ApiResponse)
def get_gene_detail(gene_name: str):
    data = svc.get_gene_detail(gene_name)
    if not data.get("sheets"):
        return ApiResponse(success=False, error=f"未找到基因: {gene_name}")
    return ApiResponse(data=data)


@router.get("/drugs", response_model=ApiResponse)
def list_drugs(
    search: str = Query("", description="Search keyword"),
    page: int = 1,
    page_size: int = 50,
):
    data = svc.get_drug_list(page=page, page_size=page_size, search=search)
    return ApiResponse(data=data)


@router.get("/immune-genes", response_model=ApiResponse)
def list_immune_genes():
    data = svc.get_immune_genes()
    return ApiResponse(data=data)


@router.get("/stats", response_model=ApiResponse)
def knowledge_stats():
    data = svc.get_stats()
    return ApiResponse(data=data)


@router.post("/reload", response_model=ApiResponse)
def reload_knowledge_bases(admin: User = Depends(require_admin)):
    """Force reload all knowledge base caches (admin only)."""
    svc.reload_all()
    return ApiResponse(data={"reloaded": True})
